### ISSUE TITLE
Passwords for new users are not masked and are stored in browser’s autocomplete cache

### ISSUE THREAD

**Issue by mvglasow** [webmin/webmin#847]
Steps to reproduce:

* Go to **System > Users and Groups**.
* Click **Create new user**.
* For the password, choose **Normal password** and enter a password in the text box.
* Fill the remaining fields and create the user.
* Start creating another user using the same procedure.
* Choose **Normal password** and, in the text box, type the first characters of the previous user’s password.

Expected behavior:

* Password is masked on entry (with only dots shown instead of actual characters)
* No autocompletion history is stored.

Behavior seen:

* All passwords entered are shown in plain text.
* Autocomplete is offered for passwords and suggests all previously entered passwords, which can easily be verified by entering the first characters of a previously supplied password or hitting the cursor-down key.

System configuration:

* Server side: Raspbian Squeeze, Webmin 1.880
* Client side: Ubuntu MATE 16.04, Firefox 58.0.2 (64-bit)

**Comment by chris001** [comment]
Docs + solution here:
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/password

**Comment by mvglasow** [comment]
Yep, that’s what needs to be changed in the code somewhere. IIRC this worked in earlier versions (around late 2016).

**Comment by jcameron** [comment]
You can make passwords hidden by going to the Module Config page of the Users and Groups module (the gear icon), and setting "Conceal plain-text password?" to "Yes"

**Comment by mvglasow** [comment]
Indeed, that produces the desired behavior. I would suggest enabling that setting by default, which currently doesn’t seem to be the case (this is a fresh install of 1.880, with no leftover settings from previous versions).

**Comment by jcameron** [comment]
Surprisingly, a lot of users want to be able to see new passwords as they enter them by default.

**Comment by mvglasow** [comment]
Well, this might easily become a security-versus-UX dilemma—but how about having users entering new passwords twice (and possibly warning about Caps Lock being on) if passwords are concealed, and making that the default? That would reduce the risk of accidentally entering a different password than intended, which is likely behind the desire to see the password entered.

**Comment by chris001** [comment]
* Letting the browser memorize the passwords for later autocomplete, is the big security issue, in this one text box for "create new password for account".

* For the function of letting the user see the passwords, best way is to add an "eyeball" icon + jquery function, to temporarily toggle the state between see the clear text and mask it.

**Comment by gnadelwartz** [comment]
IMHO `autocomplete="off"`  does not prevent a password manager to store the password, it gives a hint not to fill in an existing password. For entering new passwords `autocomplete="new-password"` may also of interest, because it hint a password manager to offer assistance for a secure new password.

I like the proposal to add an eye icon near password inputs to toggle the state temporary from hidden to visible, @iliaross  your turn :-)

**Comment by jcameron** [comment]
I wouldn't object to changing the default for this field, to become a regular password field instead of a plain text box. Not sure how that interacts with password managers though.

**Comment by chris001** [comment]
It's a special kind of field, because it's for admins to type or auto generate passwords for users, so consequently, the admin shouldn't want or need to save the users' passwords in the admin's browser keyring or password manager.

**Comment by qooob** [comment]
@gnadelwartz [`autocomplete="off"`](https://www.w3schools.com/tags/att_input_autocomplete.asp) will make sure that browser doesn't have the password stored on this field. Password manager is different story. You can let them save it, or you can not.

@chris001 And, yes, having it as non-password field is better, as it doesn't even ask for storing the password.

I think the initial request was, that the password is stored on the field, when you start typing the password again, for the new user, you see the previous entries. To prevent this, `autocomplete="off"` attr could be used on that particular field.

**Comment by gnadelwartz** [comment]
So we don't expect in this use case that the password should hidden while typing?
> Password is masked on entry (with only dots shown instead of actual characters)

**Comment by qooob** [comment]
No, and as Jamie stated it's controllable using module's config.

**Comment by chris001** [comment]
Maybe now is a good time to add a feature button, "Auto-generate a strong password", save the admin the effort of having to make a strong unique pw, every time they add a user.

**Comment by qooob** [comment]
Chris, what do you mean? We already have auto-generate password at top right of the table? Do you have it there in Authentic Theme?

**Comment by qooob** [comment]
It's automatically gets inserted to clipboard.

**Comment by qooob** [comment]
The password strength can be difined in theme's configuration globally.

**Comment by chris001** [comment]
OK @iliaross now I notice the generate password button.  Would be nice UX if it were right next to the password field, where it's used, and if it would autofill/replace the pw field contents each time it is pressed.  To save on clicking back on the field, and typing ctrl+V paste.

**Comment by qooob** [comment]
What if there are multiple password fileds on the page? It's not always possible to detect password filed, example is this module (User and Groups), where password field is plain text. Once password is inserted in the clipboard, I find it very simple to insert it and also send it to the user.

**Comment by mvglasow** [comment]
Any news on this? As of 2.202, we still have the same, potentially insecure, default behavior.

**Comment by iliajie** [comment]
I agree that we should instruct the browser not to save password fields. This has now been implemented. You can test it by applying this patch:

```
webmin patch https://github.com/webmin/webmin/commit/5931243e
```
